Use new 'goimport' tool instead of VimScript for :GoImport and :GoDrop

Carpetsmoker opened the pull request:
Instead of a large regexp-based VimScript to parse imports,
it now uses [`goimport`](https://github.com/Carpetsmoker/goimport), which parses the ast.

Fixes #1534 and deals with various other edge cases too. The behaviour is mostly the same, except in cases where it didn't make too much sense to keep the old behaviour.

For example, `:GoImport errors` will now error out if `errors` is already imported, and `:GoImport github.com/pkg/errors` will now replace any existing `errors` package instead of just adding it (this is actually the reason I started on this, because it's pretty annoying otherwise).

Carpetsmoker commented:
I maintained the current commands for compatibility btw, but we could remove `:GoImportAs` and simply make the second argument for `:GoImport` optional.

bhcleek commented:
Using the AST is going to fail when the go file is not yet parseable; e.g. if someone is in the middle of editing it and they `:GoImport time`. Consider trying to use `:GoImport` when the go file is
```
package main

func main() {
   showTime()
}

func showTime() error {
    fmt.Println(time.Now())
```

Given that `GoImport` is intended to help *while* editing, I don't think `:GoImport` should require a file that is syntactically correct...

Carpetsmoker commented:
> Using the AST is going to fail when the go file is not yet parseable

Not exactly; we pass the [`parser.ImportsOnly`](https://github.com/Carpetsmoker/goimport/blob/master/goimport.go#L159) flag, which means it will only parse the import block. There are probably some limits to what it can and can't parse, but in general it works fairly well with invalid syntax.

fatih commented:
Albeit less people use this because of goimports, I still think this is valuable. I'm +1 on this 

> For example, :GoImport errors will now error out if errors is already imported,

I don't think this should  show an error, instead it should be a noop. Just like :GoFmt, :GoAddTags, etc..

bhcleek commented:
@Carpetsmoker testing reveals `:GoImport io` will panic when run 

```
package foo

import()

func bar() {
    println("quux")
}
```

And if it will remove the package declaration if run in
```
package foo

func bar()
    println("quux")
}
```

bhcleek commented:
I haven't tested it yet, but I'd expect this solution to work with files that have build tags in them....

Carpetsmoker commented:
I fixed the panic; I'm not 100% sure what you mean with "And if it will remove the package declaration if run in"; is that related to the panic or another issue?

It works fine with build tags, as it operates on files rather than packages. I added a test for it to make sure.

bhcleek commented:
It's related to another issue. When the file does not have an import statement, I was seeing the `:Goimport`  cause the package declaration to be elided.

Carpetsmoker commented:
Oh right; that should be fixed too now.

bhcleek commented:
I found another scenario where this fails. I created two new tests: refs/pull/1559/head...bhcleek:pr-1559/more-tests

The Test_import_noemptyline one fails.

fatih commented:
@Carpetsmoker Martin, what's the state of this? Let us know how you feel about this.
